docs/style-guide.md

```diff
@@ -0,0 +1,423 @@
+* **PartialEq** is useful when a looser equality is required (e.g. comparing `NaN`s for floating point).
+* [**Ord**](https://doc.rust-lang.org/std/cmp/trait.Ord.html) and [**PartialOrd**](https://doc.rust-lang.org/std/cmp/trait.PartialOrd.html) are traits designed for ordering types.
+  * These traits define the behaviour of the comparison operators (`<`, `>`, `<=`, `>=`).
+  * Their behavior must be consistent with **Eq** and **PartialEq**.
+
+Implementing comparison traits can be problematic and should only be done for types where it is semantically meaningful. Note that Rust is very opinionated on the expected behaviour of these traits, so (for example) you must not implement a non-symmetric equality operator or a non-transitive partial ordering.
+
+One drawback to deriving the Eq trait "too early" would be if you later add some other non-Eq type to the struct. For example, **floating point types implement PartialEq but not Eq**.
+
+### No "Clever" Comparisons :: suggested
+
+The above comparison traits can be implemented for "interesting" type combination, not just on a single type. This would allow for semantically complex comparisons between user types (e.g. a mutual ordering between region codes and language codes). However this sort of API design, with implicit relations, is often subtle and easy to misuse. As the documentation for PartialEq states:
+
+> Implementations of PartialEq, PartialOrd, and Ord must agree with each other. **It's easy to accidentally make them disagree by deriving some of the traits and manually implementing others**.
+
+Note that we might choose to have non-derived orderings for some types (e.g. script sub-tags), but these should always be carefully documented.
+
+## Sized Types
+
+**Open Question**: Decide what to recommend here (if anything).
+
+Types in Rust can be **sized** or **unsized**. A sized type knows at compile time how much space is needed to hold it, and can thus have stack space allocated or be copied into a Vec or array. This improves memory locality and reduces the need for heap allocations.
+
+Perhaps more importantly, **an unsized type can only be passed by reference to other functions** and cannot be Copy-able (though they can be Clone-able). In particular any trait is, by default, unsized which makes storing traits in other structs a bit interesting (there is a [ToOwned](https://doc.rust-lang.org/std/borrow/trait.ToOwned.html) trait to help with this though).
+
+This adds weight to the idea that we should avoid traits and any other unsized types when specifying user visible types. Obviously we will accept things like `str` as parameters (these are unsized), but we aren't adding them directly to structures by reference.
+
+For example, I think that in general we should avoid returning an abstract trait to the user (intermediate traits like Iterator might be fine though since a user is expected to consume those fairly immediately).
```

Comment:
i think this will probably end up being necessary in some cases because of the pluggability. @zbraniecki can talk more about how trait objects become necessary

but i would also prefer to avoid if possible